### PR TITLE
Add tutorial

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @EPashkin

### CODE REVIEW

File: tuto/upcast_downcast.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+Since there is an inheritance system in Gtk, it's only logical to have one as well in `Gtk-rs`. Normally, most people won't need this, but it's always nice to understand how things work.
+
+## Checking if a `Widget` is another `Widget`
+
+Let's say you want to write a generic function and want to check if a `Widget` is actually a `Box`. First, let's see a bit of code:
+
+```rust
+fn is_a_box<W: IsA<gtk::Object> + IsA<gtk::Widget>>(widget: W) -> bool {
+    widget.upcast::<gtk::Widget>().downcast::<gtk::Box>().is_ok()
```
Comment (EPashkin) [inline tuto/upcast_downcast.md]: Not sure that upcast needed. Also check that consumes widget don't look good.

Comment (EPashkin) [inline tuto/upcast_downcast.md]: Oh don't read next part. Ignore it.

File: tuto/upcast_downcast.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+}
+```
+
+Then let's test it:
+
+```rust
+let button = gtk::Button::new_with_label("Click me!");
+
+is_a::<gtk::Button, gtk::Container>(button); // Returns true.
+is_a::<gtk::Button, gtk::Label>(button);     // Returns false.
```
Comment (EPashkin) [inline tuto/upcast_downcast.md]: Not sure that it works, with consuming `is_a`

File: tuto/rust_and_gtk.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+---
+layout: default
+---
+
+# Rust and `Gtk-rs`
+
+Before going any further, if you don't know Rust at all, we recommend you to read the official [Rust Book](https://doc.rust-lang.org/book/).
+
+Then we recommand you to learn a bit about [Cargo](http://doc.crates.io/index.html).
```
Comment (EPashkin) [inline tuto/rust_and_gtk.md]: recommend

File: tuto/upcast_downcast.md (hunk 1 above)
Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: I wanted to keep it simple, how would you write it?

File: tuto/upcast_downcast.md (hunk 2 above)
Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: Oh damn, I need to clone everytime. Forgot that.

File: tuto/gnome_and_rust.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * Gio
+ * Pango
+
+The goal is to provide a safe abstraction using Rust paradigms.
+
+## Objects
+
+First thing to note (it's very important!) is that the `Gtk-rs` objects can be cloned and it costs **nothing** more than copying a pointer, so basically nothing. The reason is quite simple (and I suppose you already guessed it): it's simply because `Gtk-rs` structs only contains a pointer to the corresponding Gnome objects.
+
+Now: why cloning is safe? One thing to note before going any further: it's not thread safe. If you want to handle an object in different threads, you'll have to use lockers or equivalent to avoid data racing. Otherwise, when a struct is dropped, it calls internally the [`g_object_unref`](http://gtk-rs.org/docs/gobject_sys/fn.g_object_unref.html) function and calls the [`g_object_ref`](http://gtk-rs.org/docs/gobject_sys/fn.g_object_ref.html) function when you call `clone()`.
```
Comment (EPashkin) [inline tuto/gnome_and_rust.md]: IMHO Gtk functions can't be called from non main thread

Comment (GuillaumeGomez) [inline tuto/gnome_and_rust.md]: No they're not. I'll try to word it better.

File: tuto/upcast_downcast.md (hunk 1 above)
Comment (EPashkin) [inline tuto/upcast_downcast.md]: Maybe better introduce `upcast` separately?
With some `let widget = button.upcast::<gtk::Widget>()` at end of last example

Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: Oh, good idea!

File: tuto/gnome_and_rust.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+To put it simply: Gnome handles the resources allocation/removal for us.
+
+## Trait hierarchy
+
+In Gtk, there is a widget hierarchy. In Rust, it's implemented through "traits inheritance" and enforced by the compiler at compile-time. So, what does that mean exactly? Let's take an example:
```
Comment (EPashkin) [inline tuto/gnome_and_rust.md]: As I know, we don't have direct "traits inheritance" and use `IsA` to emulate that.

Comment (GuillaumeGomez) [inline tuto/gnome_and_rust.md]: Hum... I guess you're right.

**Comment by EPashkin** [comment]
Also I have worry about all new pages will be added to header.

**Comment by GuillaumeGomez** [comment]
We'll need to check the results and I'll make quick updates in case it creates new headers.

**Comment by GuillaumeGomez** [comment]
Updated (I'll squash commits once we're good).

File: tuto/upcast_downcast.md (hunk 1 above)
Comment (EPashkin) [inline tuto/upcast_downcast.md]: Still don't see reason to have this function when `Cast` trait have `is` method.

Comment (EPashkin) [inline tuto/upcast_downcast.md]: May be change it to call `show_all` ?

Comment (EPashkin) [inline tuto/upcast_downcast.md]: .. as example of generic function.
And add `downcast` example separately

Comment (EPashkin) [inline tuto/upcast_downcast.md]: Something like replace last example with;
```
let button = ..
let widget = upcast
assert_eq!(false, widget.IsA<Label>());
if widget.isA<Button> {
  let button2 = widget.downcast::<Button>();
}
```

Comment (EPashkin) [inline tuto/upcast_downcast.md]: .. in condition `button2.set_label("new label");

Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: That's why:

```rust
assert!(button.is::<gtk::Container>());
// the trait `gtk::IsA<gtk::Button>` is not implemented for `gtk::Container`
```

`.is()` is only good to check below.

Comment (EPashkin) [inline tuto/upcast_downcast.md]: It work of compile time check that no way instance of button is container.

Comment (EPashkin) [inline tuto/upcast_downcast.md]: Oh you mean you added `is_a_box` for check working to any type.

Comment (EPashkin) [inline tuto/upcast_downcast.md]: Then maybe change function to this:
```
fn is_a_box<W: IsA<gtk::Object> + IsA<gtk::Widget> + Clone>(widget: &W) -> bool {
    widget.clone().upcast::<gtk::Widget>().is::<gtk::Box>()
}
```

Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: I thought about it. Let's do it this way then!

**Comment by GuillaumeGomez** [comment]
Updated.

File: tuto/upcast_downcast.md [hunk 6]
```diff
@@ -38,9 +38,9 @@ We don't really need our widget to be a [`Widget`](http://gtk-rs.org/docs/gtk/st
 So the point of this function is to upcast the widget to the highest widget type and then try downcasting it into the wanted object. We could make even more generic like this:
 
 ```rust
-fn is_a<W: IsA<gtk::Object> + IsA<gtk::Widget>,
-        T: IsA<gtk::Object> + IsA<gtk::Widget>>(widget: W) -> bool {
-    widget.upcast::<gtk::Widget>().downcast::<T>().is_ok()
+fn is_a<W: IsA<gtk::Object> + IsA<gtk::Widget> + Clone,
+        T: IsA<gtk::Object> + IsA<gtk::Widget>>(widget: &W) -> bool {
+    widget.clone().upcast::<gtk::Widget>().downcast::<T>().is_ok()
```
Comment (EPashkin) [inline tuto/upcast_downcast.md]: Not better without downcast?

File: tuto/upcast_downcast.md [hunk 7]
```diff
@@ -49,8 +49,8 @@ Then let's test it:
 ```rust
 let button = gtk::Button::new_with_label("Click me!");
 
-is_a::<gtk::Button, gtk::Container>(button.clone()); // Returns true.
-is_a::<gtk::Button, gtk::Label>(button.clone());     // Returns false.
+is_a::<gtk::Button, gtk::Container>(&button); // Returns true.
+is_a::<gtk::Button, gtk::Label>(&button);     // Returns false.
```
Comment (EPashkin) [inline tuto/upcast_downcast.md]: `is_a::<_, gtk::Label>(&button)` is works too
Not better show results with `assert_eq!` ?

File: tuto/upcast_downcast.md (hunk 6 above)
Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: With this way, you can make it work any time. I don't think there is a more efficient way.

File: tuto/upcast_downcast.md (hunk 7 above)
Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: Let's go for `assert!` then!

File: tuto/upcast_downcast.md [hunk 8]
```diff
@@ -49,8 +49,8 @@ Then let's test it:
 ```rust
 let button = gtk::Button::new_with_label("Click me!");
 
-is_a::<gtk::Button, gtk::Container>(&button); // Returns true.
-is_a::<gtk::Button, gtk::Label>(&button);     // Returns false.
+assert!(is_a::<_, gtk::Container>(&button)); // Won't fail.
+assert!(is_a::<_, gtk::Label>(&button));     // Will fail.
```
Comment (EPashkin) [inline tuto/upcast_downcast.md]: `!is_a` then

Comment (GuillaumeGomez) [inline tuto/upcast_downcast.md]: I'll use `assert_eq` instead then.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by EPashkin** [comment]
LGFM, Thanks.

**Comment by GuillaumeGomez** [comment]
I'll in a few minutes so I can fix in case of issues.

**Comment by EPashkin** [comment]
Strange: only http://gtk-rs.org/docs-src/ and http://gtk-rs.org/docs-src/ have many pages in headers.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Force reload (it's not the case for me).

**Comment by EPashkin** [comment]
Yes, you fixed it with last commit

**Comment by GuillaumeGomez** [comment]
I improved the footer style a bit as well. Do you see anything else I might have break as well?

**Comment by EPashkin** [comment]
I don't see any errors.

**Comment by EPashkin** [comment]
This footer is sure better

**Comment by GuillaumeGomez** [comment]
Ah you see! :D

**Pull request merged**